lib/api/ldp/ldp-operation.js

```diff
@@ -0,0 +1,300 @@
+'use strict'
+
+const url = require('url')
+const glob = require('glob')
+const HttpError = require('standard-http-error')
+const LdpTarget = require('./ldp-target')
+
+const BY_METHOD = {
+  'head': LdpHeadOperation,
+  'get': LdpGetOperation,
+  'put': LdpPutOperation,
+  'post': LdpPostOperation,
+  'patch': LdpPatchOperation,
+  'delete': LdpDeleteOperation,
+  'copy': LdpCopyOperation,
+  'options': LdpOptionsOperation
+}
+
+class LdpOperation {
+  /**
+   * @param options {object}
+   *
+   * @param options.target {LdpTarget}
+   *
+   * @param options.headers {object}
+   *
+   * @param [options.authentication=null] {Authentication} Authn object,
+   *   contains WebID string, as well as any bearer credentials/tokens. Needed
+   *   for authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.bodyStream] {Stream} Request body stream, for parsing when
+   *   needed
+   */
+  constructor (options) {
+    this.target = options.target
+    this.headers = options.headers
+    this.authentication = options.authentication
+    this.bodyStream = options.bodyStream
+  }
+
+  /**
+   * General construction method, overridden where needed (such as in
+   * LdpGetOperation).
+   *
+   * @async (to match LdpGetOperation.from() which needs to access store)
+   *
+   * @param options {object}
+   *
+   * @returns {LdpOperation}
+   */
+  static async from (options) {
+    const Operation = this
+    return new Operation(options)
+  }
+
+  writeCommonHeaders ({res, result = {}}) {
+    const { resource } = result
+
+    // set headers in common to all LDP responses
+    if (resource) { // delete operation don't return a result, for example
+      res.set('Content-Type', resource.contentType)
+      res.set('Content-Length', resource.contentLength)
+    }
+  }
+}
+
+LdpOperation.BY_METHOD = BY_METHOD
+
+/**
+ * Checks to see if target exists
+ */
+class LdpHeadOperation extends LdpOperation {
+  writeHeaders ({res, result, permissions}) {
+    this.writeCommonHeaders({res, result})
+    // write HEAD-specific headers
+  }
+
+  /**
+   * @param res {ServerResponse}
+   */
+  writeResponse ({res}) {
+    res.sendStatus(200)
```

Comment:
Is that a useful default implementation? Empty body should be 201.
(We risk creating incorrect documents, for instance, `text/html` cannot be an empty string.)

Reply:
Ok, so this is something I struggled with for a bit. (Whether to return a 200 or a 204 for a HEAD response.) And the consensus out there seems to be a 200. (I'm not sure incorrect document is a threat here, as HEAD requests by definition don't include a response body.

Reply:
No explicit mention of the status code in the RFC, but

> The HEAD method is identical to GET except that the server MUST NOT
   send a message body in the response (i.e., the response terminates at
   the end of the header section). 
— https://tools.ietf.org/html/rfc7231#section-4.3.2

So I tend to use the same status code as if it were a GET.

Reply:
Heh, yeah, the lack of explicit mention of the status code in the RFC is part of the trouble here.

> So I tend to use the same status code as if it were a GET.

Meaning what? a 200?

Reply:
200 / 204 / 404 / 410 / 500, whatever the GET would return :-)

Reply:
Ok. So I think we're on the same page. So, going back to the code, it seems that a 200 code is reasonable as a default (and that errors like 404 / 500 etc will interrupt the flow and override).

Reply:
As far as returning a 204 on a get, I agree with [this S.O. answer](https://stackoverflow.com/a/25510864) that while it's technically allowed, it's weird and somewhat nonsensical. And I'd rather avoid it. (Besides, I'm not sure how a 204 response on a GET is even possible on Solid server. So, non-issue.)